src/flp/szk.rs

```diff
@@ -57,29 +60,125 @@ pub enum SzkError {
 
 /// Contains an FLP proof share, and if joint randomness is needed, the blind
 /// used to derive it and the other party's joint randomness part.
-#[derive(Clone)]
-pub enum SzkProofShare<F: FftFriendlyFieldElement, const SEED_SIZE: usize> {
-    /// Leader's proof share is uncompressed. The first Seed is a blind, second
-    /// is a joint randomness part.
+#[derive(Debug, Clone, PartialEq)]
+pub enum SzkProofShare<F, const SEED_SIZE: usize> {
+    /// Leader's proof share is uncompressed.
     Leader {
+        /// Share of an FLP proof, as a vector of Field elements.
         uncompressed_proof_share: Vec<F>,
-        leader_blind_and_helper_joint_rand_part: Option<(Seed<SEED_SIZE>, Seed<SEED_SIZE>)>,
+        /// Set only if joint randomness is needed. The first Seed is a blind, second
+        /// is the helper's joint randomness part.
+        leader_blind_and_helper_joint_rand_part_opt: Option<(Seed<SEED_SIZE>, Seed<SEED_SIZE>)>,
     },
     /// The Helper uses one seed for both its compressed proof share and as the blind for its joint
     /// randomness.
     Helper {
+        /// The Seed that acts both as the compressed proof share and, optionally, as the blind.
         proof_share_seed_and_blind: Seed<SEED_SIZE>,
-        leader_joint_rand_part: Option<Seed<SEED_SIZE>>,
+        /// The leader's joint randomness part, if needed.
+        leader_joint_rand_part_opt: Option<Seed<SEED_SIZE>>,
     },
 }
 
+impl<F: FieldElement + Decode, const SEED_SIZE: usize> ParameterizedDecode<(bool, usize, bool)>
+    for SzkProofShare<F, SEED_SIZE>
+{
+    fn decode_with_param(
+        (is_leader, proof_len, requires_joint_rand): &(bool, usize, bool),
+        bytes: &mut Cursor<&[u8]>,
+    ) -> Result<Self, CodecError> {
+        if *is_leader {
+            Ok(SzkProofShare::Leader {
+                uncompressed_proof_share: Vec::<F>::decode_with_param(proof_len, bytes)?,
+                leader_blind_and_helper_joint_rand_part_opt: if *requires_joint_rand {
+                    Some((
+                        Seed::<SEED_SIZE>::decode(bytes)?,
+                        Seed::<SEED_SIZE>::decode(bytes)?,
+                    ))
+                } else {
+                    None
+                },
+            })
+        } else {
+            Ok(SzkProofShare::Helper {
+                proof_share_seed_and_blind: Seed::<SEED_SIZE>::decode(bytes)?,
+                leader_joint_rand_part_opt: if *requires_joint_rand {
+                    Some(Seed::<SEED_SIZE>::decode(bytes)?)
+                } else {
+                    None
+                },
```

Comment:
[then_some](https://doc.rust-lang.org/nightly/std/primitive.bool.html#method.then_some) works nicely here as well

Reply:
Since there is a side effect to `Seed::decode(bytes)?`, we can't use `then_some()`. See https://github.com/divviup/libprio-rs/pull/1087#discussion_r1681833720. If we wanted to go the combinator route, I think `requires_joint_rand.then(|| Seed::<SEED_SIZE>::decode(bytes)).transpose()?` would work.

Reply:
I'd like to use the same technique for both branches of the *is_leader if-statement, and the one above runs decode twice in a tuple, so using transpose()? instead of applying the ? operator to decode directly doesn't seem to work as intended.

Reply:
I think that may be doable with question marks both inside and outside of the closure:

```rust
flag.then(|| {
   Ok((
        Seed::<SEED_SIZE>::decode(bytes)?,
        Seed::<SEED_SIZE>::decode(bytes)?,
    ))
}).transpose()?
```

Reply:
It ended up also needing type annotations.